File: C++/GTKmm/src/widgets/tab_menu.cpp
#include "tab_menu.h"
#include "../theme.h"

TabMenu::TabMenu() : Gtk::Box(Gtk::ORIENTATION_HORIZONTAL, 0) {
    set_size_request(-1, 28);
    const struct { TabFilter f; const char* label; } tabs[] = {
        { TabFilter::All,       "All"       },
        { TabFilter::Active,    "Active"    },
        { TabFilter::Completed, "Completed" },
    };
    for (auto& t : tabs) {
        auto* btn = Gtk::make_managed<Gtk::Button>(t.label);
        btn->get_style_context()->add_class("tab-btn");
        btn->set_relief(Gtk::RELIEF_NONE);
        btn->set_focus_on_click(false);
        TabFilter f = t.f;
        btn->signal_clicked().connect([this, f](){ on_click(f); });
        btns_.push_back(btn);
        pack_start(*btn, false, false, 0);
    }
    set_active(TabFilter::All);
}

void TabMenu::set_active(TabFilter f) {
    for (int i = 0; i < 3; ++i) {
        auto ctx = btns_[i]->get_style_context();
        if (static_cast<int>(f) == i) ctx->add_class("tab-btn-active");
        else                           ctx->remove_class("tab-btn-active");
    }
}

void TabMenu::on_click(TabFilter f) {
    set_active(f);
    signal_changed.emit(f);
}

File: Python/PyGTK/widgets/tab_menu.py
import gi
gi.require_version("Gtk", "3.0")
from gi.repository import Gtk
from typing import Callable, List
from data import TabFilter
from theme import Theme

TAB_LABELS = [
    (TabFilter.ALL,       "All"),
    (TabFilter.ACTIVE,    "Active"),
    (TabFilter.COMPLETED, "Completed"),
]


class TabMenu(Gtk.Box):
    def __init__(self, on_change: Callable[[TabFilter], None]):
        super().__init__(orientation=Gtk.Orientation.HORIZONTAL, spacing=0)
        self._on_change = on_change
        self._buttons: List = []

        for f, label in TAB_LABELS:
            btn = Gtk.Button(label=label)
            btn.get_style_context().add_class("tab-btn")
            btn.connect("clicked", lambda _b, fv=f: self._click(fv))
            self._buttons.append((f, btn))
            self.pack_start(btn, False, False, 0)

        self.set_active(TabFilter.ALL)

    def set_active(self, active: TabFilter):
        for f, btn in self._buttons:
            ctx = btn.get_style_context()
            if f == active:
                ctx.add_class("tab-btn-active")
            else:
                ctx.remove_class("tab-btn-active")

    def _click(self, f: TabFilter):
        self.set_active(f)
        self._on_change(f)